File: workers/worker-rust/src/processor.rs
use crate::models::{MatchGroup, MatchItem, MatchRequest, MatchResult};
use redis::{AsyncCommands, Client};
use std::collections::{HashMap, VecDeque};
use std::env;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::time::timeout;

struct BoundedRegexCache {
    map: HashMap<String, regex::Regex>,
    order: VecDeque<String>,
    capacity: usize,
}

impl BoundedRegexCache {
    fn new(capacity: usize) -> Self {
        Self {
            map: HashMap::with_capacity(capacity),
            order: VecDeque::with_capacity(capacity),
            capacity,
        }
    }

    fn get(&mut self, key: &str) -> Option<regex::Regex> {
        self.map.get(key).cloned()
    }

    fn insert(&mut self, key: String, re: regex::Regex) {
        if !self.map.contains_key(&key) {
            if self.order.len() >= self.capacity {
                if let Some(old_key) = self.order.pop_front() {
                    self.map.remove(&old_key);
                }
            }
            self.order.push_back(key.clone());
        }
        self.map.insert(key, re);
    }
}

pub async fn listen_and_process(client: Client) {
    println!("[Worker] Rust worker listening on 'queue:rust'...");

    let mut con = client.get_async_connection().await.expect("Failed to connect to Redis");

    let timeout_ms: u64 = env::var("WORKER_EXECUTION_TIMEOUT_MS").unwrap_or_else(|_| "1000".to_string()).parse().unwrap_or(1000);
    let max_input_size: usize = env::var("WORKER_MAX_INPUT_SIZE").unwrap_or_else(|_| "10485760".to_string()).parse().unwrap_or(10485760);
    let max_matches: usize = env::var("WORKER_MAX_MATCHES").unwrap_or_else(|_| "10000".to_string()).parse().unwrap_or(10000);
    let max_groups: usize = env::var("WORKER_MAX_GROUPS").unwrap_or_else(|_| "1000".to_string()).parse().unwrap_or(1000);
    let max_json_size: usize = env::var("WORKER_MAX_JSON_SIZE").unwrap_or_else(|_| "10485760".to_string()).parse().unwrap_or(10485760);

    let cache = Arc::new(Mutex::new(BoundedRegexCache::new(1000)));

    loop {
        let result: redis::RedisResult<(String, String)> = con.brpop("queue:rust", 0.0).await;

        match result {
            Ok((_, task_json)) => {
                let mut req: MatchRequest = match serde_json::from_str(&task_json) {
                    Ok(r) => r,
                    Err(e) => {
                        handle_dlq(&client, &task_json, &format!("JSON parse error: {}", e)).await;
                        continue;
                    }
                };

                let mut con_clone = client.get_async_connection().await.unwrap();
                let mut pub_con_clone = client.get_async_connection().await.unwrap();
                let client_clone = client.clone();
                let cache_clone = Arc::clone(&cache);
                let original_json = task_json.clone();

                tokio::spawn(async move {
                    if let Some(payload_id) = &req.text_payload_id {
                        let payload_res: redis::RedisResult<String> = con_clone.get(payload_id).await;
                        match payload_res {
                            Ok(payload) => req.text = payload,
                            Err(_) => {
                                handle_dlq(&client_clone, &original_json, "Payload expired or missing from Redis").await;
                                let err_res = MatchResult {
                                    task_id: req.task_id.clone(),
                                    engine_id: req.engine_id.clone(),
                                    success: false,
                                    matches: vec![],
                                    execution_time_ms: 0.0,
                                    error: Some("Payload expired or missing from Redis".to_string()),
                                };
                                publish_result(&mut pub_con_clone, &req.task_id, err_res, max_json_size).await;
                                return;
                            }
                        }
                    }

                    let start = std::time::Instant::now();

                    let exec_future = execute_regex(&req, cache_clone, max_input_size, max_matches, max_groups);
                    let final_result = match timeout(Duration::from_millis(timeout_ms), exec_future).await {
                        Ok(res) => {
                            let mut r = res;
                            r.execution_time_ms = start.elapsed().as_micros() as f64 / 1000.0;
                            r
                        },
                        Err(_) => {
                            handle_dlq(&client_clone, &original_json, "TIMEOUT: Rust execution exceeded SLA.").await;
                            MatchResult {
                                task_id: req.task_id.clone(),
                                engine_id: req.engine_id.clone(),
                                success: false,
                                matches: vec![],
                                execution_time_ms: timeout_ms as f64,
                                error: Some(format!("TIMEOUT: Rust execution exceeded {}ms SLA.", timeout_ms)),
                            }
                        }
                    };

                    publish_result(&mut pub_con_clone, &req.task_id, final_result, max_json_size).await;
                });
            }
            Err(_) => {
                tokio::time::sleep(Duration::from_millis(100)).await;
            }
        }
    }
}

async fn handle_dlq(client: &Client, task_json: &str, error_msg: &str) {
    if let Ok(mut con) = client.get_async_connection().await {
        if let Ok(mut v) = serde_json::from_str::<serde_json::Value>(task_json) {
            if let Some(obj) = v.as_object_mut() {
                let current_attempts = obj.get("attempt_count").and_then(|a| a.as_u64()).unwrap_or(0);
                obj.insert("attempt_count".to_string(), serde_json::json!(current_attempts + 1));
                obj.insert("error_reason".to_string(), serde_json::json!(error_msg));
            }
            let _ : redis::RedisResult<()> = con.lpush("queue:rust:dead", v.to_string()).await;
        }
    }
}

async fn publish_result(con: &mut redis::aio::Connection, task_id: &str, mut result: MatchResult, max_json_size: usize) {
    let mut res_json = serde_json::to_string(&result).unwrap_or_default();

    if res_json.len() > max_json_size {
        result.success = false;
        result.matches = vec![];
        result.error = Some(format!("Output JSON exceeds maximum allowed size of {} bytes.", max_json_size));
        res_json = serde_json::to_string(&result).unwrap_or_default();
    }

    let key = format!("result:{}", task_id);
    let _ : redis::RedisResult<()> = con.set_ex(&key, res_json, 60).await;
    let _ : redis::RedisResult<()> = con.publish(&key, "ready").await;
}

async fn execute_regex(
    req: &MatchRequest,
    cache: Arc<Mutex<BoundedRegexCache>>,
    max_input_size: usize,
    max_matches: usize,
    max_groups: usize
) -> MatchResult {

    if req.text.len() > max_input_size {
        return MatchResult {
            task_id: req.task_id.clone(),
            engine_id: req.engine_id.clone(),
            success: false,
            matches: vec![],
            execution_time_ms: 0.0,
            error: Some(format!("Input text exceeds maximum allowed size of {} bytes.", max_input_size)),
        };
    }

    let mut case_insensitive = false;
    let mut multi_line = false;
    let mut dot_matches_new_line = false;
    let mut crlf = false;
    let mut swap_greed = false;
    let mut unicode = true;
    let mut ignore_whitespace = false;

    let mut flag_str = String::new();

    for flag in &req.flags {
        flag_str.push_str(flag);
        match flag.as_str() {
            "i" => case_insensitive = true,
            "m" => multi_line = true,
            "s" => dot_matches_new_line = true,
            "R" => crlf = true,
            "U" => swap_greed = true,
            "u" => unicode = true,
            "-u" => unicode = false,
            "x" => ignore_whitespace = true,
            _ => {}
        }
    }

    let cache_key = format!("{}|{}", flag_str, req.regex);

    let re_opt = {
        let mut c = cache.lock().unwrap();
        c.get(&cache_key)
    };

    let re = match re_opt {
        Some(r) => r,
        None => {
            let builder = regex::RegexBuilder::new(&req.regex)
                .case_insensitive(case_insensitive)
                .multi_line(multi_line)
                .dot_matches_new_line(dot_matches_new_line)
                .crlf(crlf)
                .swap_greed(swap_greed)
                .unicode(unicode)
                .ignore_whitespace(ignore_whitespace)
                .size_limit(10 * (1 << 20))
                .build();

            match builder {
                Ok(r) => {
                    cache.lock().unwrap().insert(cache_key, r.clone());
                    r
                }
                Err(e) => return MatchResult {
                    task_id: req.task_id.clone(),
                    engine_id: req.engine_id.clone(),
                    success: false,
                    matches: vec![],
                    execution_time_ms: 0.0,
                    error: Some(format!("Compilation failed: {}", e)),
                }
            }
        }
    };

    let mut match_items = Vec::new();
    let mut match_id = 0;

    for caps in re.captures_iter(&req.text) {
        if match_id >= max_matches {
            return MatchResult {
                task_id: req.task_id.clone(),
                engine_id: req.engine_id.clone(),
                success: false,
                matches: vec![],
                execution_time_ms: 0.0,
                error: Some(format!("Exceeded maximum allowed matches ({}).", max_matches)),
            };
        }

        if let Some(full_match) = caps.get(0) {
            let mut groups = Vec::new();

            for (i, name_opt) in re.capture_names().enumerate().skip(1) {
                if groups.len() >= max_groups {
                    return MatchResult {
                        task_id: req.task_id.clone(),
                        engine_id: req.engine_id.clone(),
                        success: false,
                        matches: vec![],
                        execution_time_ms: 0.0,
                        error: Some(format!("Exceeded maximum allowed groups per match ({}).", max_groups)),
                    };
                }

                if let Some(m) = caps.get(i) {
                    groups.push(MatchGroup {
                        group_id: i,
                        name: name_opt.map(|s| s.to_string()),
                        content: m.as_str().to_string(),
                        start: m.start(),
                        end: m.end(),
                    });
                }
            }

            match_items.push(MatchItem {
                match_id,
                full_match: full_match.as_str().to_string(),
                start: full_match.start(),
                end: full_match.end(),
                groups,
            });
            match_id += 1;
        }
    }

    MatchResult {
        task_id: req.task_id.clone(),
        engine_id: req.engine_id.clone(),
        success: true,
        matches: match_items,
        execution_time_ms: 0.0,
        error: None,
    }
}

File: workers/worker-go/processor.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

type RegexCache struct {
	sync.RWMutex
	m    map[string]*regexp.Regexp
	keys []string
}

func (c *RegexCache) Get(key string) *regexp.Regexp {
	c.RLock()
	defer c.RUnlock()
	return c.m[key]
}

func (c *RegexCache) Set(key string, r *regexp.Regexp) {
	c.Lock()
	defer c.Unlock()
	if _, exists := c.m[key]; !exists {
		c.keys = append(c.keys, key)
		if len(c.keys) > 1000 {
			delete(c.m, c.keys[0])
			c.keys = c.keys[1:]
		}
	}
	c.m[key] = r
}

var regexCache = &RegexCache{m: make(map[string]*regexp.Regexp)}

func getEnvInt(key string, fallback int) int {
	if val, ok := os.LookupEnv(key); ok {
		if i, err := strconv.Atoi(val); err == nil {
			return i
		}
	}
	return fallback
}

func handleDlq(ctx context.Context, client *redis.Client, taskJson string, errorMsg string) {
	var taskMap map[string]interface{}
	if err := json.Unmarshal([]byte(taskJson), &taskMap); err == nil {
		attemptCount := 0.0
		if val, ok := taskMap["attempt_count"].(float64); ok {
			attemptCount = val
		}
		taskMap["attempt_count"] = attemptCount + 1
		taskMap["error_reason"] = errorMsg

		if modifiedJson, err := json.Marshal(taskMap); err == nil {
			client.LPush(ctx, "queue:go:dead", string(modifiedJson))
		}
	}
}

func listenAndProcess(client *redis.Client) {
	fmt.Println("[Worker] Go worker listening on 'queue:go'...")
	bgCtx := context.Background()

	timeoutMs := getEnvInt("WORKER_EXECUTION_TIMEOUT_MS", 1000)
	maxInputSize := getEnvInt("WORKER_MAX_INPUT_SIZE", 10485760)
	maxMatches := getEnvInt("WORKER_MAX_MATCHES", 10000)
	maxGroups := getEnvInt("WORKER_MAX_GROUPS", 1000)
	maxJsonSize := getEnvInt("WORKER_MAX_JSON_SIZE", 10485760)

	for {
		result, err := client.BRPop(bgCtx, 0, "queue:go").Result()
		if err != nil {
			time.Sleep(100 * time.Millisecond)
			continue
		}

		go func(taskJson string) {
			var req MatchRequest
			if err := json.Unmarshal([]byte(taskJson), &req); err != nil {
				handleDlq(bgCtx, client, taskJson, fmt.Sprintf("JSON parse error: %v", err))
				return
			}

			// CLAIM-CHECK: Extract text_payload_id dynamically
			var claimCheck struct {
				TextPayloadID *string `json:"text_payload_id"`
			}
			_ = json.Unmarshal([]byte(taskJson), &claimCheck)

			if claimCheck.TextPayloadID != nil && *claimCheck.TextPayloadID != "" {
				payload, err := client.Get(bgCtx, *claimCheck.TextPayloadID).Result()
				if err != nil {
					errStr := "Payload expired or missing from Redis"
					handleDlq(bgCtx, client, taskJson, errStr)

					res := MatchResult{
						TaskID:          req.TaskID,
						EngineID:        req.EngineID,
						Success:         false,
						Matches:         []MatchItem{},
						ExecutionTimeMs: 0.0,
						Error:           &errStr,
					}
					resJson, _ := json.Marshal(res)
					client.SetEx(bgCtx, "result:"+req.TaskID, string(resJson), 60*time.Second)
					client.Publish(bgCtx, "result:"+req.TaskID, "ready")
					return
				}
				req.Text = payload
			}

			start := time.Now()
			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutMs)*time.Millisecond)
			defer cancel()

			resultChan := make(chan MatchResult, 1)

			go func() {
				res := MatchResult{
					TaskID:   req.TaskID,
					EngineID: req.EngineID,
				}

				if len(req.Text) > maxInputSize {
					execTime := float64(time.Since(start).Microseconds()) / 1000.0
					errStr := fmt.Sprintf("Input text exceeds maximum allowed size of %d bytes.", maxInputSize)
					res.Success = false
					res.Matches = []MatchItem{}
					res.ExecutionTimeMs = execTime
					res.Error = &errStr
					resultChan <- res
					return
				}

				inlineFlagPrefix, flagErr := buildGoRegexInlineFlagPrefix(req.Flags)
				if flagErr != nil {
					execTime := float64(time.Since(start).Microseconds()) / 1000.0
					errStr := flagErr.Error()
					res.Success = false
					res.Matches = []MatchItem{}
					res.ExecutionTimeMs = execTime
					res.Error = &errStr
					resultChan <- res
					return
				}

				fullRegex := inlineFlagPrefix + req.Regex

				re := regexCache.Get(fullRegex)
				if re == nil {
					var err error
					re, err = regexp.Compile(fullRegex)
					if err != nil {
						execTime := float64(time.Since(start).Microseconds()) / 1000.0
						errStr := err.Error()
						res.Success = false
						res.Matches = []MatchItem{}
						res.ExecutionTimeMs = execTime
						res.Error = &errStr
						resultChan <- res
						return
					}
					regexCache.Set(fullRegex, re)
				}

				submatches := re.FindAllStringSubmatchIndex(req.Text, maxMatches+1)
				names := re.SubexpNames()

				var matchItems []MatchItem

				for matchId, matchIdxs := range submatches {
					if matchId >= maxMatches {
						execTime := float64(time.Since(start).Microseconds()) / 1000.0
						errStr := fmt.Sprintf("Exceeded maximum allowed matches (%d).", maxMatches)
						res.Success = false
						res.Matches = []MatchItem{}
						res.ExecutionTimeMs = execTime
						res.Error = &errStr
						resultChan <- res
						return
					}

					var groups []MatchGroup
					fullMatch := req.Text[matchIdxs[0]:matchIdxs[1]]

					for i := 1; i < len(matchIdxs)/2; i++ {
						if len(groups) >= maxGroups {
							execTime := float64(time.Since(start).Microseconds()) / 1000.0
							errStr := fmt.Sprintf("Exceeded maximum allowed groups per match (%d).", maxGroups)
							res.Success = false
							res.Matches = []MatchItem{}
							res.ExecutionTimeMs = execTime
							res.Error = &errStr
							resultChan <- res
							return
						}

						groupStart := matchIdxs[2*i]
						groupEnd := matchIdxs[2*i+1]

						if groupStart != -1 && groupEnd != -1 {
							var namePtr *string
							if i < len(names) && names[i] != "" {
								n := names[i]
								namePtr = &n
							}

							groups = append(groups, MatchGroup{
								GroupID: i,
								Name:    namePtr,
								Content: req.Text[groupStart:groupEnd],
								Start:   groupStart,
								End:     groupEnd,
							})
						}
					}

					if groups == nil {
						groups = []MatchGroup{}
					}

					matchItems = append(matchItems, MatchItem{
						MatchID:   matchId,
						FullMatch: fullMatch,
						Start:     matchIdxs[0],
						End:       matchIdxs[1],
						Groups:    groups,
					})
				}

				if matchItems == nil {
					matchItems = []MatchItem{}
				}

				execTime := float64(time.Since(start).Microseconds()) / 1000.0
				res.Success = true
				res.Matches = matchItems
				res.ExecutionTimeMs = execTime
				resultChan <- res
			}()

			var finalRes MatchResult
			select {
			case finalRes = <-resultChan:
				// Execution completed
			case <-ctx.Done():
				// SLA Timeout
				handleDlq(bgCtx, client, taskJson, "TIMEOUT: Go execution exceeded SLA.")
				execTime := float64(time.Since(start).Microseconds()) / 1000.0
				errStr := fmt.Sprintf("TIMEOUT: Go execution exceeded %dms SLA.", timeoutMs)
				finalRes = MatchResult{
					TaskID:          req.TaskID,
					EngineID:        req.EngineID,
					Success:         false,
					Matches:         []MatchItem{},
					ExecutionTimeMs: execTime,
					Error:           &errStr,
				}
			}

			resJson, _ := json.Marshal(finalRes)

			if len(resJson) > maxJsonSize {
				errStr := fmt.Sprintf("Output JSON exceeds maximum allowed size of %d bytes.", maxJsonSize)
				finalRes = MatchResult{
					TaskID:          req.TaskID,
					EngineID:        req.EngineID,
					Success:         false,
					Matches:         []MatchItem{},
					ExecutionTimeMs: finalRes.ExecutionTimeMs,
					Error:           &errStr,
				}
				resJson, _ = json.Marshal(finalRes)
			}

			client.SetEx(bgCtx, "result:"+req.TaskID, string(resJson), 60*time.Second)
			client.Publish(bgCtx, "result:"+req.TaskID, "ready")

		}(result[1])
	}
}